[mlir][tosa] Update Tensorflow to match TOSA v1.0 specification (part 3)

wonjeon opened the pull request:
We've been pushing TOSA v1.0 LLVM patches to upstream. This PR includes commits to keep the following operators to be aligned with LLVM:

[mlir][tosa] Change 'shape' of RESHAPE from attribute to input shape
https://github.com/llvm/llvm-project/pull/125789
[mlir][tosa] Make Convolution Zero Points Inputs
https://github.com/llvm/llvm-project/pull/122939
[mlir][tosa] Remove Quantization Attribute
https://github.com/llvm/llvm-project/pull/125479
[mlir][tosa] Change ClampOp's min/max attributes
https://github.com/llvm/llvm-project/pull/125197
[mlir][tosa] Make TOSA RESIZE's scale, offset, border as Input
https://github.com/llvm/llvm-project/pull/124956

This patch including:

[mlir][tosa] Change 'shape' attribute of RESHAPE operator to become an input

Signed-off-by: Won Jeon <won.jeon@arm.com>
Change-Id: I938503349f38b64db5e77a01c3a7b2bb33e8f041

[mlir][tosa] Switch zero point of convolutions to input variable type

Change-Id: Ifd3f24779886377735e6694f716f23c5137f9138

[Tosa] Refactor QuantizationAttr

changes due to removal of quantization attr in TOSA dialect and due to name changes in while_loop region names

Signed-off-by: Tai Ly <tai.ly@arm.com>
Change-Id: I09533bffcd8e2179505c7e11e1320b673266585d

[Tosa] ClampOp attributes changes

This patch implements changes required by Tosa ClampOp's new min_val/max_val attributes

Signed-off-by: Tai Ly <tai.ly@arm.com>
Change-Id: I25ba0d077fa44d4c384ab094a6070a4743383414

[TOSA] Calculate unknown reshape dimension when input is static

This commit updates the reshape legalization to calculate static shape and output type when a static input shape is provided and only one dimension is unknown.

Change-Id: I0843549b47131b0852fbf375f00846b1fcbe8bc6
Signed-off-by: Luke Hutton <luke.hutton@arm.com>

[TOSA] Numerical mismatch on tfl.transpose_conv layer

* Legalization now handles cases where the layer has a bias

Change-Id: Ie3ba38644d1cf8e5d6f71271e8bb6f1b5636f406

[mlir][tosa] Change resize attrs to inputs

This patch implements changes required by Tosa resize op's
scale/offset/border changing from attributes to inputs.

Signed-off-by: Tai Ly <[tai.ly](http://tai.ly/)@[arm.com](http://arm.com/)>
Change-Id: I9a4319ac53298c25568fc651e249528b9a9477fc

wonjeon commented:
Local testing successfully done:

INFO: Build completed successfully, 27 total actions
//tensorflow/compiler/mlir/tosa/tests:convert-tfl-uint8.mlir.test        PASSED in 0.3s
//tensorflow/compiler/mlir/tosa/tests:convert_metadata.mlir.test         PASSED in 0.2s
//tensorflow/compiler/mlir/tosa/tests:fuse-bias-tf.mlir.test             PASSED in 0.3s
//tensorflow/compiler/mlir/tosa/tests:lower-complex-types.mlir.test      PASSED in 0.3s
//tensorflow/compiler/mlir/tosa/tests:multi_add.mlir.test                PASSED in 0.3s
//tensorflow/compiler/mlir/tosa/tests:retain_call_once_funcs.mlir.test   PASSED in 0.2s
//tensorflow/compiler/mlir/tosa/tests:strip-quant-types.mlir.test        PASSED in 0.2s
//tensorflow/compiler/mlir/tosa/tests:strip_metadata.mlir.test           PASSED in 0.3s
//tensorflow/compiler/mlir/tosa/tests:tf-tfl-to-tosa-pipeline.mlir.test  PASSED in 0.3s
//tensorflow/compiler/mlir/tosa/tests:tf-to-tosa-pipeline.mlir.test      PASSED in 0.4s
//tensorflow/compiler/mlir/tosa/tests:tf-unequal-ranks.mlir.test         PASSED in 0.4s
//tensorflow/compiler/mlir/tosa/tests:tfl-to-tosa-dequantize_softmax.mlir.test PASSED in 0.2s
//tensorflow/compiler/mlir/tosa/tests:tfl-to-tosa-pipeline-filtered.mlir.test PASSED in 0.3s
//tensorflow/compiler/mlir/tosa/tests:tfl-to-tosa-pipeline.mlir.test     PASSED in 2.9s
//tensorflow/compiler/mlir/tosa/tests:tfl-to-tosa-stateful.mlir.test     PASSED in 0.4s
//tensorflow/compiler/mlir/tosa/tests:tfl-unequal-ranks.mlir.test        PASSED in 0.3s
//tensorflow/compiler/mlir/tosa/tests:verify_fully_converted.mlir.test   PASSED in 0.3s

Executed 17 out of 17 tests: 17 tests pass.

sjarus commented:
Internal CI with bazel test passing. 

Assigning to @GeorgeARM .

wonjeon commented:
Tested and confirmed that this patch works with the following LLVM integration:
https://github.com/tensorflow/tensorflow/commit/01197637cb4eee7da2f1645f7367268ee44efa87

wonjeon commented:
Tested and confirmed that this patch works with the LLVM integration, https://github.com/tensorflow/tensorflow/commit/4d8657f0c13540570c65bb8fc9a17b9d66755f6f.

lhutton1 reviewed (commented):
Overall LGTM though it's quite hard to review the individual changes when merged like this. Perhaps this can be made easier by pushing multiple individual patches such that they appear separately in the "Commits" tab, these changes can then be squashed when merging into main?

wonjeon committed:
[mlir][tosa] Change 'shape' attribute of RESHAPE operator to become an input

including minor change from:
Slice the input of kernel based ops to the actual used size

Change-Id: Ifebe0d1b3459300df0fa2edc9ba24a867caec3d3

Signed-off-by: Won Jeon <won.jeon@arm.com>
Change-Id: I938503349f38b64db5e77a01c3a7b2bb33e8f041

wonjeon committed:
[Tosa] Refactor QuantizationAttr

changes due to removal of quantization attr in TOSA dialect
and due to name changes in while_loop region names

Signed-off-by: Tai Ly <tai.ly@arm.com>
Change-Id: I09533bffcd8e2179505c7e11e1320b673266585d

wonjeon committed:
[mlir][tosa] ClampOp attributes changes

This patch implements changes required by Tosa ClampOp's
new min_val/max_val attributes

including clamp_max update code from:

commit 04055fa510522af659aa56bac3b4796961131546
Author: Thibaut Goetghebuer-Planchon <thibaut.goetghebuer-planchon@arm.com>
Date:   Thu Sep 21 17:22:14 2023 +0000

    [TOSA] During quantized ReLU legalization, limit the clamp_max attribute to the max value of the quantized type

    Change-Id: I781229be0eb86ecb3cf1a305ede98ad630e5bcfd

Signed-off-by: Tai Ly <tai.ly@arm.com>
Change-Id: I25ba0d077fa44d4c384ab094a6070a4743383414

wonjeon committed:
[TOSA] Calculate unknown reshape dimension when input is static

This commit updates the reshape legalization to calculate static
shape and output type when a static input shape is provided and
only one dimension is unknown.

Change-Id: I0843549b47131b0852fbf375f00846b1fcbe8bc6
Signed-off-by: Luke Hutton <luke.hutton@arm.com>

wonjeon committed:
[TOSA] Numerical mismatch on tfl.transpose_conv layer

* Legalization now handles cases where the layer has a bias

Author: Tom Allsop <tom.allsop@arm.com>
Change-Id: Ie3ba38644d1cf8e5d6f71271e8bb6f1b5636f406

wonjeon committed:
[mlir][tosa] Change resize attrs to inputs

This patch implements changes required by Tosa resize op's
scale/offset/border changing from attributes to inputs.

Signed-off-by: Tai Ly <tai.ly@arm.com>
Change-Id: I9a4319ac53298c25568fc651e249528b9a9477fc

wonjeon committed:
[mlir][tosa] Update LIT tests

Combination of test file updates from the following commits:
* [mlir][tosa] Change 'shape' attribute of RESHAPE operator to become an input
* [mlir][tosa] Switch zero point of convolutions to input variable type
* [Tosa] Refactor QuantizationAttr
* [TOSA] During quantized ReLU legalization, limit the clamp_max attribute to the ma
x value of the quantized type
* [mlir][tosa] ClampOp attributes changes
* [TOSA] Calculate unknown reshape dimension when input is static
* [TOSA] Numerical mismatch on tfl.transpose_conv layer
* [mlir][tosa] Change resize attrs to inputs

Co-authored-by: Tai Ly <tai.ly@arm.com>
Co-authored-by: Thibaut Goetghebuer-Planchon <thibaut.goetghebuer-planchon@arm.com>
Co-authored-by: Luke Hutton <luke.hutton@arm.com>
Co-authored-by: Tom Allsop <tom.allsop@arm.com>

Signed-off-by: Won Jeon <won.jeon@arm.com>
Change-Id: Ia5731e659d262c74374e8326d49beccf6a60032e

wonjeon commented:
@lhutton1 , thanks for your comments. Changed the PR so that it contains multiple individual patches plus the test file changes. Hopefully this helps for your review.

wonjeon commented:
Local testing successfully done:

INFO: Analyzed 33 targets (0 packages loaded, 33384 targets configured).
INFO: Found 16 targets and 17 test targets...
INFO: Elapsed time: 2.061s, Critical Path: 0.25s
INFO: 2 processes: 2 local.
INFO: Build completed successfully, 2 total actions
//tensorflow/compiler/mlir/tosa/tests:convert-tfl-uint8.mlir.test (cached) PASSED in 0.2s
//tensorflow/compiler/mlir/tosa/tests:convert_metadata.mlir.test (cached) PASSED in 0.3s
//tensorflow/compiler/mlir/tosa/tests:fuse-bias-tf.mlir.test    (cached) PASSED in 0.3s
//tensorflow/compiler/mlir/tosa/tests:lower-complex-types.mlir.test (cached) PASSED in 0.2s
//tensorflow/compiler/mlir/tosa/tests:multi_add.mlir.test       (cached) PASSED in 0.2s
//tensorflow/compiler/mlir/tosa/tests:retain_call_once_funcs.mlir.test (cached) PASSED in 0.2s
//tensorflow/compiler/mlir/tosa/tests:strip-quant-types.mlir.test (cached) PASSED in 0.2s
//tensorflow/compiler/mlir/tosa/tests:strip_metadata.mlir.test  (cached) PASSED in 0.2s
//tensorflow/compiler/mlir/tosa/tests:tf-tfl-to-tosa-pipeline.mlir.test (cached) PASSED in 0.3s
//tensorflow/compiler/mlir/tosa/tests:tf-to-tosa-pipeline.mlir.test (cached) PASSED in 0.5s
//tensorflow/compiler/mlir/tosa/tests:tf-unequal-ranks.mlir.test (cached) PASSED in 0.2s
//tensorflow/compiler/mlir/tosa/tests:tfl-to-tosa-dequantize_softmax.mlir.test (cached) PASSED in 0.2s
//tensorflow/compiler/mlir/tosa/tests:tfl-to-tosa-pipeline-filtered.mlir.test (cached) PASSED in 0.3s
//tensorflow/compiler/mlir/tosa/tests:tfl-to-tosa-stateful.mlir.test (cached) PASSED in 0.4s
//tensorflow/compiler/mlir/tosa/tests:tfl-unequal-ranks.mlir.test (cached) PASSED in 0.2s
//tensorflow/compiler/mlir/tosa/tests:verify_fully_converted.mlir.test (cached) PASSED in 0.3s
//tensorflow/compiler/mlir/tosa/tests:tfl-to-tosa-pipeline.mlir.test     PASSED in 0.1s

Executed 1 out of 17 tests: 17 tests pass.

lhutton1 reviewed (approved):
Thanks for breaking into logical commits @wonjeon!

sjarus reviewed (approved):
Approving to trigger precommit.
